### PR TITLE
Use glob in file selector

### PR DESCRIPTION
**Pull request by staronj**

### CODE REVIEW

File: src/dune/file_selector.ml [hunk 1]
```diff
@@ -1,31 +1,66 @@
...
+    function
+    | Predicate predicate ->
+      Variant ("predicate", [ Predicate.to_dyn predicate ])
+    | Glob glob -> Variant ("glob", [ Glob.to_dyn glob ])
+
+  let encode =
+    let open Dune_lang.Encoder in
+    function
+    | Predicate predicate ->
+      record [ ("predicate", Predicate.encode predicate) ]
```
Comment (aalekseyev) [inline src/dune/file_selector.ml]: `record` results in an extra layer of parentheses. Strangely, `variant` function doesn't exist and instead in `dep.ml` we're using `pair`.
Maybe you can add `val variant : string -> sexp list -> sexp` and use that?

**Comment by aalekseyev** [comment]
The patch looks all right to me, but on the high level I'm not sure if it will be able to serve its purpose.

The purpose is to be able to serialize and desiralize a (subset of) `Dep.t` values so we can talk to the child process about extra dependencies.
I think ideally we have a separate type that captures such a subset, at which point we might find that we don't need this change in its current form.

What do you think?

**Comment by diml** [comment]
What's the story for versioning the serialised values? We should expect that the client side library used by commands will not always match the version of dune. For instance, the client side library might be vendored in a mono-repository.

**Comment by aalekseyev** [comment]
I think I was wrong about the purpose of this change at first: this seems to be for computing dependency digests, not for communicating with the child process.

The problem @diml mentions should be relevant for `Dune_action.Protocol` sexp conversions. I guess we should slap a version number at the top-level of the Sexp and that should be enough?

**Comment by diml** [comment]
> The problem @diml mentions should be relevant for Dune_action.Protocol sexp conversions. I guess we should slap a version number at the top-level of the Sexp and that should be enough?

That seems fine yes. BTW, we should make sure that a newer dune is able to communicate with an older library.

**Comment by aalekseyev** [comment]
Sounds like we're going with a separate file for marshallable deps instead of this change. Closing.

**Pull request closed without merging**